### PR TITLE
"Lifetimes with Serde" assignment

### PR DESCRIPTION
**Pull request by listochkin**
This assignment starts with JSON deserializing into a struct with own
`String` and gradually move towards use of `Cow`.

Lifetimes show up once, but the initial code is crafted in such a way
that you have to act on your reference potentially outliving
the underlying owned `String`.

The assignment is very quick to complete with most lines of code either
put directly into the text itself or linked from the description.
However, there's a potential of a good discussion, and to facilitate it
I've put a lift of questions at the end.

### CODE REVIEW

File: assignments/serde-lifetimes.adoc [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+#[derive(Debug)]
+struct BlogPost {
+    id: u32,
+
+    title: String,
+}
+
+fn main() -> Result<(), serde_json::Error> {
+    let post: BlogPost = {
+        let data = fetch_data();
```
Comment (amanjeev) [inline assignments/serde-lifetimes.adoc:45]: @listochkin I am confused here. wouldn't this be dropped at the end of the block?

File: assignments/serde-lifetimes.adoc [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    {
+        "color": "#ff00ff"
+    }
+"##
+----
+====
+
+== Tasks
+
+=== 1. Serialization and Deserialization using `String` s
```
Comment (amanjeev) [inline assignments/serde-lifetimes.adoc:77]: Perhaps it is a good idea to mention that you want the participants to do all `todo`s here in the Task 1 itself.

File: assignments/serde-lifetimes.adoc [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+Turns out Serde creates an owned copy of the data anyway. To avoid that use link:https://serde.rs/field-attrs.html#borrow[`#[serde(borrow)\]`] macro.
+
+Get to the point when your program prints: `is borrowed = true`
+
+== Questions after this exercise
+
+ 1. Why is there a need for separate `serde` and `serde-json` crates?
+ 2. Why do we need to put a lifetime annotation for a field of '&str' type?
+ 3. What about `Cow` type?
```
Comment (amanjeev) [inline assignments/serde-lifetimes.adoc:146]: Maybe reframe this question more clearly? I am not sure what you are asking about.

File: assignments/serde-lifetimes.adoc (hunk 1 above)
Comment (listochkin) [reply assignments/serde-lifetimes.adoc:45]: That's the point. Step 2: Deserialization using `str` has the following text:

> Use lifetime annotations and change the code inside main if necessary.
>
> Q: If you made a change, why?

The idea is to let the participants stumble on your structure potentially outliving the underlying data storage, and that switch to reference-basedstructures can cause potential re-design of your code.

File: assignments/serde-lifetimes.adoc (hunk 2 above)
Comment (listochkin) [reply assignments/serde-lifetimes.adoc:77]: Isn't it what the very next line says?

Comment (amanjeev) [reply assignments/serde-lifetimes.adoc:77]: Ah yea, I think I got confused that should this step one should do all `todo()`s or not. I guess its not needed. My bad.

**Review by amanjeev — COMMENTED** [review]
Thankyou!

**Review by amanjeev — APPROVED** [review]

**Pull request merged**